src/mesh/parallel_tree.jl

```diff
@@ -0,0 +1,833 @@
+  println(io, "t.capacity = $(t.capacity)")
+  println(io, "t.length = $(t.length)")
+  println(io, "t.dummy = $(t.dummy)")
+  println(io, "t.center_level_0 = $(t.center_level_0)")
+  println(io, "t.length_level_0 = $(t.length_level_0)")
+  println(io, '*'^20)
+end
+
+# Type traits to obtain dimension
+@inline Base.ndims(t::Type{ParallelTree{NDIMS}}) where NDIMS = NDIMS
+@inline Base.ndims(t::ParallelTree) = ndims(typeof(t))
+
+
+# Auxiliary methods to allow semantic queries on the tree
+# Check whether cell has parent cell
+has_parent(t::ParallelTree, cell_id::Int) = t.parent_ids[cell_id] > 0
+
+# Count number of children for a given cell
+n_children(t::ParallelTree, cell_id::Int) = count(x -> (x > 0), @view t.child_ids[:, cell_id])
+
+# Check whether cell has any child cell
+has_children(t::ParallelTree, cell_id::Int) = n_children(t, cell_id) > 0
+
+# Check whether cell is leaf cell
+is_leaf(t::ParallelTree, cell_id::Int) = !has_children(t, cell_id)
+
+# Check whether cell has specific child cell
+has_child(t::ParallelTree, cell_id::Int, child::Int) = t.child_ids[child, cell_id] > 0
+
+# Check if cell has a neighbor at the same refinement level in the given direction
+has_neighbor(t::ParallelTree, cell_id::Int, direction::Int) = t.neighbor_ids[direction, cell_id] > 0
+
+# Check if cell is own cell, i.e., belongs to this MPI rank
+is_own_cell(t::ParallelTree, cell_id) = t.mpi_ranks[cell_id] == mpi_rank()
+
+# Check if cell has a coarse neighbor, i.e., with one refinement level lower
+function has_coarse_neighbor(t::ParallelTree, cell_id::Int, direction::Int)
+  return has_parent(t, cell_id) && has_neighbor(t, t.parent_ids[cell_id], direction)
+end
+
+# Check if cell has any neighbor (same-level or lower-level)
+function has_any_neighbor(t::ParallelTree, cell_id::Int, direction::Int)
+  return has_neighbor(t, cell_id, direction) || has_coarse_neighbor(t, cell_id, direction)
+end
+
+# Return cell length for a given level
+length_at_level(t::ParallelTree, level::Int) = t.length_level_0 / 2^level
+
+# Return cell length for a given cell
+length_at_cell(t::ParallelTree, cell_id::Int) = length_at_level(t, t.levels[cell_id])
+
+# Return minimum level of any leaf cell
+minimum_level(t::ParallelTree) = minimum(t.levels[leaf_cells(t)])
+
+# Return maximum level of any leaf cell
+maximum_level(t::ParallelTree) = maximum(t.levels[leaf_cells(t)])
+
+# Check if tree is periodic
+isperiodic(t::ParallelTree) = all(t.periodicity)
+isperiodic(t::ParallelTree, dimension) = t.periodicity[dimension]
+
+
+# Auxiliary methods for often-required calculations
+# Number of potential child cells
+n_children_per_cell(::ParallelTree{NDIMS}) where NDIMS = 2^NDIMS
+# n_children_per_cell(dims::Integer) = 2^dims
+
+# Number of directions
+#
+# Directions are indicated by numbers from 1 to 2*ndims:
+# 1 -> -x
+# 2 -> +x
+# 3 -> -y
+# 4 -> +y
+# 5 -> -z
+# 6 -> +z
+n_directions(::ParallelTree{NDIMS}) where NDIMS = 2 * NDIMS
```

Comment:
Shall we try to reduce the amount of code duplication by making methods like this dispatch on `AbstractTree{NDIMS}` and move them to another file?

Reply:
See https://github.com/trixi-framework/Trixi.jl/pull/167#discussion_r495718977. I am closing this here, but feel free to unresolve if necessary.